### PR TITLE
test: use job metadata to avoid test flakes

### LINKED ISSUE #151: tests.system.reader.test_reader: test_snapshot[v1] failed
This test failed!

To configure my behavior, see [the Flaky Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/master/packages/flakybot).

If I'm commenting on this issue too often, add the `flakybot: quiet` label and
I will stop commenting.

---

commit: f1bc4128e099e83d5a15408bd6fd2129c2151301
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/679dd086-80e7-40bf-a5e8-dfd64c0c6d62), [Sponge](http://sponge2/679dd086-80e7-40bf-a5e8-dfd64c0c6d62)
status: failed
<details><summary>Test output</summary><br><pre>client_and_types = (<google.cloud.bigquery_storage_v1.BigQueryReadClient object at 0x7f8b42171fd0>, <module 'google.cloud.bigquery_storage_v1.gapic_types' from '/tmpfs/src/github/python-bigquery-storage/google/cloud/bigquery_storage_v1/gapic_types.py'>)
project_id = 'precise-truck-742'
table_with_data_ref = 'projects/precise-truck-742/datasets/bq_storage_system_tests_42335d05_507f_48e0_8c32_033823270111/tables/users'
bq_client = <google.cloud.bigquery.client.Client object at 0x7f8b42149ca0>

    def test_snapshot(client_and_types, project_id, table_with_data_ref, bq_client):
        client, types = client_and_types
        before_new_data = dt.datetime.now(tz=dt.timezone.utc)
    
        # load additional data into the table
        new_data = [
            {"first_name": "NewGuyFoo", "last_name": "Smith", "age": 46},
            {"first_name": "NewGuyBar", "last_name": "Jones", "age": 30},
        ]
    
        destination = _to_bq_table_ref(table_with_data_ref)
        bq_client.load_table_from_json(new_data, destination).result()
    
        # read data using the timestamp before the additional data load
    
        read_session = types.ReadSession()
        read_session.table = table_with_data_ref
        read_session.table_modifiers.snapshot_time = before_new_data
        read_session.data_format = types.DataFormat.AVRO
    
        session = client.create_read_session(
            request={
                "parent": "projects/{}".format(project_id),
                "read_session": read_session,
                "max_stream_count": 1,
            }
        )
        stream = session.streams[0].name
    
        rows = list(client.read_rows(stream).rows(session))
    
        # verify that only the data before the timestamp was returned
>       assert len(rows) == 5  # all initial records
E       assert 0 == 5
E        +  where 0 = len([])

tests/system/reader/test_reader.py:212: AssertionError</pre></details>

### PR DESCRIPTION
**Pull request by tswast**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-storage/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #151  🦕

### CODE REVIEW

**Comment by tswast** [comment]
@crwilcox I am trying with a `sleep` since I suspect clock sync issues. Would you prefer I use `pytest-repeat` / `flaky` as you suggest in go/handling-flakes-examples?

**Comment by crwilcox** [comment]
@tswast I think this is reasonable. If you see the go example this is the strategy there, though we don't wait 10s. Instead it does an exponential backoff up to 3 times. That way most cases don't take that much longer.

File: tests/system/reader/test_reader.py [hunk 1]
```diff
@@ -179,7 +180,12 @@ def test_column_selection_read(
 
 def test_snapshot(client_and_types, project_id, table_with_data_ref, bq_client):
     client, types = client_and_types
+    # Sleep for a moment to give use some wiggle room in case the BigQuery
+    # snapshot time and our times are out-of-sync.
+    # https://github.com/googleapis/python-bigquery-storage/issues/151
+    time.sleep(10)
```
Comment (crwilcox) [inline tests/system/reader/test_reader.py:186]: What is the point of the initial sleep? it seems the second is to ensure that the snapshot time requested is in the past?

**Review by shollyman — COMMENTED** [review]
If you have a reference to the table metadata as well as the job statistics from the load you're doing, you may be able to adapt the logic from the [table delete/undelete snippet](https://github.com/GoogleCloudPlatform/golang-samples/blob/master/bigquery/snippets/table/bigquery_undelete_table.go#L47) and avoid the wasted time sleeping.

File: tests/system/reader/test_reader.py (hunk 1 above)
Comment (tswast) [reply tests/system/reader/test_reader.py:186]: The flake is actually that there are 0 rows, so the snapshot time is too _early_ in that case.

That said, Seth makes a good point that I can use the times from the table metadata.

**Comment by tswast** [comment]
@shollyman Good point! I've updated the test to use the times from the job metadata.

**Review by shollyman — APPROVED** [review]

**Comment by tswast** [comment]
This test flaked again. https://source.cloud.google.com/results/invocations/f821c579-a7fb-45a7-9110-dad4dd830842/targets/github%2Fpython-bigquery-storage/tests

Looks like we should add a `_retry_403` like we have in other BQ tests to deal with rate limiting.

**Pull request merged**